use_cases/eluc/prescriptors/heuristics/heuristics.py

```diff
@@ -0,0 +1,127 @@
+    
+    def _reco_heuristic(self, pct: float, context_df: pd.DataFrame):
+        """
+        Takes evenly from all columns and adds to best col.
+        Removes land_use * (total_change / changeable_land) so we remove proportionally
+        rather than truly evenly.
+        """
+        adjusted = context_df.copy()
+        adjusted["scaled_change"] = pct * adjusted[constants.LAND_USE_COLS].sum(axis=1)
+        adjusted["row_sum"] = adjusted[self.presc_cols].sum(axis=1)
+        to_change = adjusted["row_sum"] > 0
+        adjusted["max_change"] = adjusted[["scaled_change", "row_sum"]].min(axis=1)
+        # Reduce all columns by even amount
+        for col in self.presc_cols:
+            adjusted.loc[to_change, col] -= adjusted.loc[to_change, col] * adjusted.loc[to_change, "max_change"] / adjusted.loc[to_change, "row_sum"]
+        # Increase best column by max change
+        adjusted.loc[to_change, self.best_col] = adjusted.loc[to_change, self.best_col] + adjusted.loc[to_change, "max_change"]
+        adjusted = adjusted.drop(["scaled_change", "row_sum", "max_change"], axis=1)
+        return adjusted
+
+class PerfectHeuristic(HeuristicPrescriptor):
+    """
+    Implementation of HeuristicPrescriptor that does an informed land use prescription 
+    based on linear regression coefficients.
+    """
+    def __init__(self, coefs: list[float], predictor: Predictor):
+        """
+        We save and sort the columns by highest coefficient i.e. most emissions.
+        Separate the best column according to the coefficients to add to.
+        """
+        super().__init__(predictor)
+        assert len(coefs) == len(constants.RECO_COLS)
+        # Sort columns by coefficient
+        reco_cols = [col for col in constants.RECO_COLS]
+        zipped = zip(reco_cols, coefs)
+        sorted_zip = sorted(zipped, key=lambda x: x[1], reverse=True)
+        self.reco_cols, _ = zip(*sorted_zip)
+        self.reco_cols = list(self.reco_cols)
+
+        # Take best column and remove from reco cols
+        self.best_col = self.reco_cols[-1]
+        self.reco_cols.pop()
+
+    def _reco_heuristic(self, pct: float, context_df: pd.DataFrame) -> pd.DataFrame:
+        """
+        Perfect prescription algorithm:
+            1. Subtract up to scaled change starting from worst column.
+            2. Add to best column however much was removed.
+        It is done in a vectorized way to massively speed up computation.
+        """
+        adjusted = context_df.copy()
+
+        adjusted["scaled_change"] = pct * adjusted[constants.LAND_USE_COLS].sum(axis=1)
+        adjusted["presc_sum"] = adjusted[self.reco_cols].sum(axis=1)
+        adjusted["to_change"] = adjusted[["scaled_change", "presc_sum"]].min(axis=1)
+
+        for col in self.reco_cols:
+            change = adjusted[[col, "to_change"]].min(axis=1)
+            adjusted[col] -= change
+            adjusted["to_change"] -= change
```

Comment:
This took me a really long time to figure out but I'm glad it ended up being so elegant vectorized rather than by row. It also led me to find there was a slight error in my previous perfect heuristic computation. It was giving a change of NaN to all rows where the land use summed to 0, shifting all the average changes slightly to the right. This made our heuristic hypervolumes very slightly worse.

Reply:
Nice find 👍